benchmarks/rack_framework/benchmark.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+require "harness"
+
+Dir.chdir(__dir__)
+use_gemfile
+
+require "rack"
+#require "rack/builder"
+
+# Start a server
+app = Rack::Builder.new do
+  #use Rack::CommonLogger
+  map "/ok" do
+    run lambda { |env| [200, {'content-type' => 'text/plain'}, ['OK']] }
+  end
+end
+
+#run app
+
+run_benchmark(10) do
+  10_000.times do
+    # The app may mutate `env`, so we need to create one every time.
+    env = Rack::MockRequest::env_for("http://localhost/ok")
```

Comment:
I'll try that. In general the answer would be "no" since the body is a StringIO that stores a read position (though maybe with a deep dup?). But we don't use the body here, so maybe it will just work.

Reply:
Interesting. A lot quicker, but less well-optimised.

```
itr #89: 111ms
itr #90: 111ms
RSS: 35.0MiB
Average of last 75, non-warmup iters: 112ms
Total time spent benchmarking: 21s

interp: ruby 3.3.0dev (2023-07-09T09:45:51Z :detached: 9dcdffb8bf) [x86_64-linux]
yjit: ruby 3.3.0dev (2023-07-09T09:45:51Z :detached: 9dcdffb8bf) +YJIT [x86_64-linux]

--------------  -----------  ----------  ---------  ----------  ------------  -----------
bench           interp (ms)  stddev (%)  yjit (ms)  stddev (%)  yjit 1st itr  interp/yjit
rack_framework  135.0        0.9         112.1      1.1         1.13          1.20
--------------  -----------  ----------  ---------  ----------  ------------  -----------
Legend:
- yjit 1st itr: ratio of interp/yjit time for the first benchmarking iteration.
- interp/yjit: ratio of interp/yjit time. Higher is better for yjit. Above 1 represents a speedup.
```

Reply:
Oddly, it *raises* the % in YJIT. But probably a lot more useful this way. Thank you!

```
Writing file /home/ubuntu/ym/yjit-bench/data/results-ruby-3.3.0-2023-07-10-170124.json
***YJIT: Printing YJIT statistics on exit***
method call exit reasons:
      block_arg:     10,451 (99.9%)
    interrupted:          7 ( 0.1%)
invokeblock exit reasons:
      proc:        608 (86.9%)
    symbol:         92 (13.1%)
invokesuper exit reasons:
    block:         67 (100.0%)
leave exit reasons:
    interp_return:     96,489 (100.0%)
     se_interrupt:          9 ( 0.0%)
getblockparamproxy exit reasons:
    (all relevant counters are zero)
getinstancevariable exit reasons:
    (all relevant counters are zero)
setinstancevariable exit reasons:
    (all relevant counters are zero)
definedivar exit reasons:
    (all relevant counters are zero)
opt_aref exit reasons:
    (all relevant counters are zero)
expandarray exit reasons:
    (all relevant counters are zero)
opt_getinlinecache exit reasons:
    (all relevant counters are zero)
invalidation reasons:
    constant_state_bump:         53 (88.3%)
       constant_ic_fill:          6 (10.0%)
          method_lookup:          1 ( 1.7%)
num_send:                    629,694
num_send_known_class:         56,802 ( 9.0%)
num_send_polymorphic:            754 ( 0.1%)
num_send_x86_rel32:            1,922
num_send_x86_reg:                  2
iseq_stack_too_large:              0
iseq_too_long:                     0
temp_reg_opnd:                 9,827
temp_mem_opnd:                 7,096
temp_spill:                    6,063
bindings_allocations:            165
bindings_set:                      0
compiled_iseq_entry:             151
compiled_iseq_count:             187
compiled_blockid_count:        1,697
compiled_block_count:          2,073
versions_per_block:            1.222
compiled_branch_count:         3,549
block_next_count:              1,891
defer_count:                     727
defer_empty_count:               165
branch_insn_count:               185
branch_known_count:               31 (16.8%)
freed_iseq_count:                 20
invalidation_count:               60
constant_state_bumps:              0
get_ivar_max_depth:                0
inline_code_size:            189,586
outlined_code_size:          187,928
code_region_size:            393,216
freed_code_size:                   0
yjit_alloc_size:             866,961
live_context_size:            95,033
live_context_count:            3,277
live_page_count:                  24
freed_page_count:                  0
code_gc_count:                     0
num_gc_obj_refs:               1,321
object_shape_count:              473
side_exit_count:              21,472
total_exit_count:            117,961
total_insns_count:         5,271,073
vm_insns_count:              687,193
yjit_insns_count:          4,605,352
ratio_in_yjit:                 87.0%
avg_len_in_yjit:                38.9
Top-9 most frequent exit ops (100.0% of exits):
                      send:     10,451 (48.7%)
        getblockparamproxy:      9,972 (46.4%)
               invokeblock:        700 ( 3.3%)
               invokesuper:        299 ( 1.4%)
    opt_send_without_block:         32 ( 0.1%)
                     leave:          9 ( 0.0%)
      opt_getconstant_path:          6 ( 0.0%)
             setlocal_WC_0:          2 ( 0.0%)
                   opt_mod:          1 ( 0.0%)
```